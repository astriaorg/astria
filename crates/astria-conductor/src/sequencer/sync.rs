use astria_sequencer_types::SequencerBlockData;
use color_eyre::eyre::{
    self,
    WrapErr as _,
};
use deadpool::managed::{
    Pool,
    PoolError,
};
use futures::stream::FuturesOrdered;
use sequencer_client::tendermint::block::Height;
use tokio::select;
use tracing::{
    error,
    info,
    instrument,
    warn,
};

use crate::client_provider::{
    self,
    ClientProvider,
};

#[derive(Debug, thiserror::Error)]
enum Error {
    #[error("failed requesting a client from the pool")]
    Pool(#[from] PoolError<client_provider::Error>),
    #[error("getting a block from sequencer failed")]
    Request(#[from] sequencer_client::extension_trait::Error),
}

#[instrument(name = "sync sequencer", skip(client_pool, executor))]
pub(super) async fn run(
    start: Height,
    end: Height,
    client_pool: Pool<ClientProvider>,
    executor: crate::executor::Sender,
) -> eyre::Result<()> {
    use futures::{
        FutureExt as _,
        StreamExt as _,
    };

    let start: u32 = start
        .value()
        .try_into()
        .wrap_err("start cometbft height overflowed u32")?;
    let end: u32 = end
        .value()
        .try_into()
        .wrap_err("end cometbft height overflowed u32")?;
    let mut height_stream = futures::stream::iter(start..end);
    let mut block_stream = FuturesOrdered::new();

    'sync: loop {
        select!(
            // The condition on block_stream relies on the pool size being currently set to 50.
            // This ensures that no more than 20 requests to the sequencer are active at the same time.
            // Leaving some objects in the pool is important so that failed blocks can be rescheduled
            // in the match arm below.
            Some(height) = height_stream.next(), if block_stream.len() <= 20 => {
                let pool = client_pool.clone();
                block_stream.push_back(async move {
                    get_client_then_block(pool, height).await
                }.map(move |res| (height, res)).boxed());
            }

            Some((height, res)) = block_stream.next() => {
                match res {
                    Err(Error::Request(error)) => {
                        let error = &error as &(dyn std::error::Error + 'static);
                        warn!(height, error, "failed getting sequencer block; rescheduling");
                        let pool = client_pool.clone();
                        block_stream.push_front(async move {
                            get_client_then_block(pool, height).await
                        }.map(move |res| (height, res)).boxed());
                    }

                    Err(Error::Pool(e)) => {
                        let error = &e as &(dyn std::error::Error + 'static);
                        error!(height, error, "failed getting a client from the pool; aborting sync");
                        break 'sync Err(e).wrap_err("failed getting a client from the pool");
                    }

                    Ok(block) => {
                        let block = Box::new(block);
                        if let Err(e) = executor.send(crate::executor::ExecutorCommand::FromSequencer { block }) {
                            let error = &e as &(dyn std::error::Error + 'static);
                            error!(height, error, "failed forwarding block to executor; aborting async");
                            break 'sync Err(e).wrap_err("failed forwarding block to executor");
                        }
                    }
                }
            }

            else => {
                info!("sync finished");
                break 'sync Ok(())
            }
        )
    }
}

async fn get_client_then_block(
    pool: Pool<ClientProvider>,
    height: u32,
) -> Result<SequencerBlockData, Error> {
    use sequencer_client::SequencerClientExt as _;

    let client = pool.get().await?;
    let block = client.sequencer_block(height).await?;
    Ok(block)
}
