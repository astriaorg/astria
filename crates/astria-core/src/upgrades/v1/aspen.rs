use std::sync::Arc;

use borsh::BorshSerialize;

use super::{
    Change,
    ChangeName,
    UpgradeName,
};
use crate::{
    generated::upgrades::v1::{
        aspen::{
            PriceFeedChange as RawPriceFeedChange,
            ValidatorUpdateActionChange as RawValidatorUpdateActionChange,
        },
        Aspen as RawAspen,
        BaseUpgradeInfo as RawBaseUpgradeInfo,
    },
    protocol::genesis::v1::{
        PriceFeedGenesis,
        PriceFeedGenesisError,
    },
    Protobuf,
};

#[derive(Clone, Debug)]
pub struct Aspen {
    activation_height: u64,
    app_version: u64,
    price_feed_change: PriceFeedChange,
    validator_update_action_change: ValidatorUpdateActionChange,
}

impl Aspen {
    pub const NAME: UpgradeName = UpgradeName::new("aspen");

    #[must_use]
    pub fn activation_height(&self) -> u64 {
        self.activation_height
    }

    #[must_use]
    pub fn app_version(&self) -> u64 {
        self.app_version
    }

    #[must_use]
    pub fn price_feed_change(&self) -> &PriceFeedChange {
        &self.price_feed_change
    }

    #[must_use]
    pub fn validator_update_action_change(&self) -> &ValidatorUpdateActionChange {
        &self.validator_update_action_change
    }

    pub fn changes(&self) -> impl Iterator<Item = &'_ dyn Change> {
        Some(&self.price_feed_change as &dyn Change)
            .into_iter()
            .chain(Some(&self.validator_update_action_change as &dyn Change))
    }
}

impl Protobuf for Aspen {
    type Error = Error;
    type Raw = RawAspen;

    fn try_from_raw_ref(raw: &Self::Raw) -> Result<Self, Self::Error> {
        let RawBaseUpgradeInfo {
            activation_height,
            app_version,
        } = *raw.base_info.as_ref().ok_or_else(Error::no_base_info)?;

        let price_feed_change = raw
            .price_feed_change
            .as_ref()
            .ok_or_else(Error::no_price_feed_change)?;

        let genesis = price_feed_change
            .genesis
            .as_ref()
            .ok_or_else(Error::no_price_feed_genesis)
            .and_then(|raw_genesis| {
                PriceFeedGenesis::try_from_raw_ref(raw_genesis).map_err(Error::price_feed_genesis)
            })?;

        if raw.validator_update_action_change.is_none() {
            return Err(Error::no_validator_update_action_change());
        }

        let price_feed_change = PriceFeedChange {
            activation_height,
            app_version,
            genesis: Arc::new(genesis),
        };

        let validator_update_action_change = ValidatorUpdateActionChange {
            activation_height,
            app_version,
        };

        Ok(Self {
            activation_height,
            app_version,
            price_feed_change,
            validator_update_action_change,
        })
    }

    fn to_raw(&self) -> Self::Raw {
        let base_info = Some(RawBaseUpgradeInfo {
            activation_height: self.activation_height,
            app_version: self.app_version,
        });
        let price_feed_change = Some(RawPriceFeedChange {
            genesis: Some(self.price_feed_change.genesis.to_raw()),
        });
        RawAspen {
            base_info,
            price_feed_change,
            validator_update_action_change: Some(RawValidatorUpdateActionChange {}),
        }
    }
}

/// This change enables vote extensions and starts to provide price feed data from the price feed
///  (if enabled) via the vote extensions.
///
/// The vote extensions are enabled in the block immediately after `activation_height`, meaning the
/// price feed data is available no earlier than two blocks after `activation_height`.
#[derive(Clone, Debug, BorshSerialize)]
pub struct PriceFeedChange {
    activation_height: u64,
    app_version: u64,
    genesis: Arc<PriceFeedGenesis>,
}

impl PriceFeedChange {
    pub const NAME: ChangeName = ChangeName::new("price_feed_change");

    #[must_use]
    pub fn genesis(&self) -> &Arc<PriceFeedGenesis> {
        &self.genesis
    }
}

impl Change for PriceFeedChange {
    fn name(&self) -> ChangeName {
        Self::NAME.clone()
    }

    fn activation_height(&self) -> u64 {
        self.activation_height
    }

    fn app_version(&self) -> u64 {
        self.app_version
    }
}

/// This change introduces new sequencer `Action`s to support updating the validator set.
#[derive(Clone, Debug, BorshSerialize)]
pub struct ValidatorUpdateActionChange {
    activation_height: u64,
    app_version: u64,
}

impl ValidatorUpdateActionChange {
    pub const NAME: ChangeName = ChangeName::new("validator_update_action_change");
}

impl Change for ValidatorUpdateActionChange {
    fn name(&self) -> ChangeName {
        Self::NAME.clone()
    }

    fn activation_height(&self) -> u64 {
        self.activation_height
    }

    fn app_version(&self) -> u64 {
        self.app_version
    }
}

/// An error when transforming a [`RawPriceFeedUpgrade`] into a [`PriceFeedChange`].
#[derive(Debug, thiserror::Error)]
#[error(transparent)]
pub struct Error(ErrorKind);

impl Error {
    fn no_base_info() -> Self {
        Self(ErrorKind::FieldNotSet("base_info"))
    }

    fn no_price_feed_change() -> Self {
        Self(ErrorKind::FieldNotSet("price_feed_change"))
    }

    fn no_validator_update_action_change() -> Self {
        Self(ErrorKind::FieldNotSet("validator_update_action_change"))
    }

    fn no_price_feed_genesis() -> Self {
        Self(ErrorKind::FieldNotSet("price_feed_change.genesis"))
    }

    fn price_feed_genesis(source: PriceFeedGenesisError) -> Self {
        Self(ErrorKind::PriceFeedGenesis {
            source,
        })
    }
}

#[derive(Debug, thiserror::Error)]
enum ErrorKind {
    #[error("`{0}` field was not set")]
    FieldNotSet(&'static str),
    #[error("`price_feed_change.genesis` field was invalid")]
    PriceFeedGenesis { source: PriceFeedGenesisError },
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{
        protocol::test_utils::dummy_price_feed_genesis,
        upgrades::v1::change::DeterministicSerialize,
    };

    #[test]
    fn serialized_price_feed_change_should_not_change() {
        let price_feed_change = PriceFeedChange {
            activation_height: 10,
            app_version: 2,
            genesis: Arc::new(dummy_price_feed_genesis()),
        };
        let serialized_price_feed_change = hex::encode(price_feed_change.to_vec());
        insta::assert_snapshot!("price_feed_change", serialized_price_feed_change);
    }

    #[test]
    fn serialized_validator_update_action_change_should_not_change() {
        let validator_update_action_change = ValidatorUpdateActionChange {
            activation_height: 10,
            app_version: 2,
        };
        let serialized_validator_update_action_change =
            hex::encode(validator_update_action_change.to_vec());
        insta::assert_snapshot!(
            "validator_update_action_change",
            serialized_validator_update_action_change
        );
    }
}
