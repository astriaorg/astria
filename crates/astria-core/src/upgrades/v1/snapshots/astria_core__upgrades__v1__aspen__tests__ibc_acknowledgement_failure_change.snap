---
source: crates/astria-core/src/upgrades/v1/aspen.rs
assertion_line: 319
expression: serialized_ibc_acknowledgement_failure_change
---
0a000000000000000200000000000000
